rollup/sync_service.go

```diff
@@ -0,0 +1,763 @@
+	}
+	if err != nil {
+		return fmt.Errorf("Cannot fetch transaction %s: %w", ethlog.TxHash.Hex(), err)
+	}
+	if pending {
+		return fmt.Errorf("Transaction %s unexpectedly in mempool", ethlog.TxHash.Hex())
+	}
+
+	cd := appendSequencerBatchCallData{}
+	err = cd.Decode(bytes.NewReader(tx.Data()))
+	if err != nil {
+		return fmt.Errorf("Cannot decode sequencer batch appended calldata: %w", err)
+	}
+
+	for i, element := range cd.ChainElements {
+		var tx *types.Transaction
+		index := event.TotalElements.Uint64() - (uint64(i) + event.NumQueueElements.Uint64())
+		// Sequencer transaction
+		if element.IsSequenced {
+			// Different types of transactions can be included in the canonical
+			// transaction chain. The first byte specifies what kind of
+			// transaction it is. Parse the data emitted from the log and then
+			// build the tx to be played against the evm based on the type
+			ctcTx := CTCTransaction{}
+			err = ctcTx.Decode(element.TxData)
+			if err != nil {
+				return fmt.Errorf("Cannot deserialize txdata: %w", err)
+			}
+
+			switch ctcTx.typ {
+			case CTCTransactionTypeEOA:
+				nonce := uint64(0)
+				to := common.Address{} // sequencerDecompressionAddress
+				gasLimit := uint64(0)  // max gas limit
+
+				tx = types.NewTransaction(nonce, to, big.NewInt(0), gasLimit, big.NewInt(0), element.TxData, nil, nil, types.QueueOriginSequencer, types.SighashEIP155)
+				tx, err = s.signTransaction(tx)
+				if err != nil {
+					return fmt.Errorf("Cannot add signature to create eoa tx: %w", err)
+				}
+			case CTCTransactionTypeEIP155:
+				eip155, ok := ctcTx.tx.(*CTCTxEIP155)
+				if !ok {
+					return fmt.Errorf("Unexpected type when parsing ctc tx eip155: %T", ctcTx.tx)
+				}
+
+				// TODO: double check the l1TxOrigin
+				nonce, gasLimit := uint64(eip155.nonce), uint64(eip155.gasLimit)
+				to, l1TxOrigin := eip155.target, common.Address{}
+				gasPrice := new(big.Int).SetUint64(uint64(eip155.gasPrice))
+				data := eip155.data
+				tx = types.NewTransaction(nonce, to, big.NewInt(0), gasLimit, gasPrice, data, &l1TxOrigin, nil, types.QueueOriginSequencer, types.SighashEIP155)
+
+				// `WithSignature` accepts:
+				// r || s || v where v is normalized to 0 or 1
+				tx, err = tx.WithSignature(s.signer, eip155.Signature[:])
+				if err != nil {
+					return fmt.Errorf("Cannot add signature to eip155 tx: %w", err)
+				}
+			default:
+				// This should never happen
+				return fmt.Errorf("Unknown tx type: %x", element.TxData)
+			}
+		} else {
+			// Queue transaction
+			rtx, ok := s.txCache.Load(index)
+			if !ok {
+				log.Error("Cannot find transaction in transaction cache", "index", index)
+				continue
+			}
+
+			tx, err = s.signTransaction(rtx.tx)
+			if err != nil {
+				log.Error("Sequencer Batch Append sign queue transaction failed", "message", err.Error())
+				continue
+			}
+			s.bc.SetCurrentTimestamp(rtx.timestamp.Unix())
+		}
+
+		err = s.maybeReorgAndApplyTx(index, tx)
```

Comment:
need to do a timestamp comparison, the batch context chain element has the timestamp